### PR TITLE
ENH: Adds PeekableIterator

### PR DESCRIPTION
**Pull request by llllllllll**
Here is one that I was using in (quasiquotes)[https://github.com/llllllllll/quasiquotes/blob/master/quasiquotes/codec/tokenizer.py#L290] when working on a custom tokenize to look ahead at the next tokens to detirmine which branch to execute without pulling all the tokens out of the stream. I am also using it in some data loading code to check on data types before writing stuff.

I realize there is already a `peek` and that this sort of does a lot of the same things. I am not sure how you feel about having both but I am offering it if you are iterested.

### CODE REVIEW

**Comment by mrocklin** [comment]
I'm mildly against this sort of thing.  Being a library of "mostly just functions for iterators" has been `toolz`' niche for a while.  This is the sort of thing I would expect to see in a project like `fn.py` which tends to be a bit more fully featured.

**Comment by llllllllll** [comment]
No worries, I can close this issue. I wasn' sure if this was generally applicable to this library but I figured I could see what you all thought.

**Comment by mrocklin** [comment]
I would also wait to see what @eriknw thinks.  I'm pretty much always happy to defer to his opinion.

**Comment by llllllllll** [comment]
The test failure is an except block and a py2 compat line in python 3. I will wait on updating until we know if it is wanted.

**Comment by eriknw** [comment]
My thoughts mirror @mrocklin's.  If I ever needed to peek at multiple elements, I would make a function for it.  What is the advantage of making it a class?

**Comment by llllllllll** [comment]
In the project I am using it in I wanted to be able to pass the peekable stream around. Sort of like `tee`, you cannot use the underlying stream after you use the peekable version or it will cause problems so I thought it was easist to abstract this state into an object. I understand that this is out of the scope of this project though

**Comment by eriknw** [comment]
Well, I'm also in favor of avoiding `tee`.  I'd like to discuss this with a few `toolz` users in person next week, okay?

File: toolz/itertoolz.py [hunk 1]
```diff
@@ -896,3 +897,70 @@ def peek(seq):
...
+    from the stream.
+
+    Parameters
+    ----------
+    stream : iterator
+        The iterator to wrap.
+    """
+    def __init__(self, stream):
+        self._stream = iter(stream)
+        self._peeked = Queue()
```
Comment (eriknw) [inline toolz/itertoolz.py]: I would recommend using `collections.deque`.  `Queue` is intended for multithreading.

**Pull request closed without merging**